### PR TITLE
First draft of custom notations

### PR DESCRIPTION
**Pull request by dan-simon**
This allows for numbers of symbols/letters other than 26, allows for multicharacter symbols/letters, and allows for spaces between mantissa or exponent or between symbols in the exponent. It doesn't allow for reversal of symbols or no mantissa (those would be easy but seem niche) or other bases (that would be harder but still possible). Also letter/cancer internal inheritance structure is changed.

I don't think I dealt with the this.letters property of custom/letter/cancer notations in the best way; as it is, outside code can change it. Not sure what to do instead but surely there's some better solution.

This has more design issues than most PRs to this repo (which are generally bugfixes or additions of a fairly well-specified notation). If you (whoever's reading this, whether an AD dev or a random person) have some extra capability you'd like custom notations to support, mention it here and I'll at least take it into consideration. Having said that, it isn't intended that the custom notation builder can make any notation whatsoever, so too-niche capacities will probably not be added (though I might make a more flexible custom community notation if there's enough demand). If you think this PR makes custom notations too difficult to customize or too complicated, say that here too.

### CODE REVIEW

**Comment by dan-simon** [comment]
Code quality error, r128. (What do I have to do to get rid of you?) Basically, as far as I see, there are two options for some parameters with default values and either gives a quality error (one because the default parameter doesn't have a typedef, the other because its type is trivially inferable). Razen, if you have time, please tell me whether you prefer the last commit or not ("Try fixing last code quality error") so I can revert it if necessary.

File: src/custom.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import { EngineeringNotation } from "./engineering";
+import Decimal from "break_infinity.js/break_infinity";
+import { toEngineering } from "./utils";
+
+export class CustomNotation extends EngineeringNotation {
+  private letters: string | string[];
+  private mantissaExponentSeparator: string;
+  private separator: string;
+  
+  constructor(letters: string | string[], mantissaExponentSeparator="", separator="") {
```
Comment (razenpok) [inline src/custom.ts]: - Validate parameters before `super` call
- Validate input parameters, not `this.parameter`
- Use more formal error messages, these are usually easier to get right
- Spaces

```suggestion
  constructor(letters: string | string[], mantissaExponentSeparator = "", separator = "") {
    if (letters.length < 2) {
      throw new Error("The supplied letter sequence must contain at least 2 letters");
    super();
    this.letters = letters;
    this.mantissaExponentSeparator = mantissaExponentSeparator;
    this.separator = separator;
  }
```

**Review by razenpok — CHANGES_REQUESTED** [review]

**Comment by razenpok** [comment]
And now dist directory is not in sync with the source code. I suggest you only commit dist directory when you are making a release. Otherwise it will end up in inconsistencies or/and in PR conflicts.

**Comment by dan-simon** [comment]
Ugh, you're right. Not sure what the right thing to do now is. I think you're right about not committing dist being wise going forward (thus having a convention of dist not being in sync in the source code), as long as whenever there's a bugfix and no known remaining bugs we make a release (so that the github.io page stays updated, not that it matters for this PR). For now I guess we can keep this branch's dist as it currently is and hope no one else makes a PR soon?

**Review by razenpok — APPROVED** [review]

**Comment by razenpok** [comment]
Ok, let's call it a day and merge it as is (I see you've updated dist and docs as well). Keeping dist and docs from the last release only seems to me as the best solution. I'll also take a look at Codacy's code quality issue sometime later.

**Comment by dan-simon** [comment]
Yeah, keeping dist and docs from the last release only and not committing it seems good to me going forward, but as I said in the commit description, for this branch we were in trouble already because new dist and docs were already committed in a previous commit and so any new PR would conflict anyway. For future PRs that's not an issue and I won't commit dist and docs at all.

**Pull request merged**